### ISSUE TITLE
Alignment of code with recent C++ standards

### ISSUE THREAD

**Issue by royfalk** [vegastrike/Vega-Strike-Engine-Source#481]
With the introduction of C++20, I think we need to set a policy of adoption of new standards.
We are currently at C++11.

The way I see it, we have the following options:

1. Stay at C++11 indefinitely. 
2. Latest and greatest - adopt the latest as soon as it is stable.
3. Some in between policy.

**Pros of conservatism**
* Code features are familiar to many.
* Great compiler support and stability.
* Ability to compile on older systems.

**Cons of conservatism**
* Potentially nicer/cleaner/more elegant code.
* Opportunity for people to experience newer C++, not always available at work. May lure developers.

I suggest we discuss this here or in a dedicated room at gitter and then vote.

**Comment by danielrh** [comment]
C++11 may have better support on OSX and windows to bring in more players -- the only thing really valuable is from C++17 (option, variant classes)---the rest (such as std::span) can come in via various header-only libraries that will defer to C++20, if available (eg https://github.com/martinmoene/span-lite/tree/master/include/nonstd )

**Comment by BenjamenMeyer** [comment]
The main driver for which version we support is the platforms we support; the secondary driver is what we want to support.
So #40 certainly comes into play here. I believe all current platforms support C++11; possibly C++14 (not sure there).

That said, C++11 has both good things (standardized types) and bad things (auto for one - especially with older codebases since they changed the meaning of the term from a variable scope to a variable type with no interluding version that didn't define it at all).

I don't think we should stay with a specific version indefinitely (e.g option 1); but neither can we always adopt the latest (option 2). So we need something along the lines of option 3.

Personally, I would define a policy based on the code base and then adopt features of newer versions as we find them beneficial and as they are supported by platforms we support.

For instance, with respect to `auto` I would define a very strict policy on its usage where as an option like option 3 would allow it near anywhere, similar with the ISO C++ guidelines.

**Comment by BenjamenMeyer** [comment]
NOTE: I'll request that any feature that we adopt as part of option 3 should have some clear examples listed in our coding style to help with understanding its proper usage.

**Comment by evertvorster** [comment]
Option 3 also seems the right option for me. 
Are the various standards not backwards compatible?

Unfortunately I do not have much time these days to contribute to Vega Strike.

**Comment by BenjamenMeyer** [comment]
@evertvorster sadly there are some backwards incompatible things - but that was mainly introduced in C++11 with the complete change in meaning of the `auto` keyword.

**Comment by royfalk** [comment]
We are already at C++11, so if auto was used in the old way, we have a bug.

As for me, the most interesting thing is modules (C++20), which can potentially reduce compilation time. Though that's less relevant ever since I bought a 16 core machine.

I did however write more elegant code somewhere using C++17 and reverted because our CICD rejected it.

**Comment by stephengtuggy** [comment]
Option 3 seems like the right one for me as well. Something like "support the latest and greatest that we reasonably can, while still maintaining our commitments to support older platforms."

There isn't enough time in life to document and decide on every new feature individually. I suggest that we just document the things we are **not** going to adopt right away. `auto` can certainly be one of those.

**Comment by stephengtuggy** [comment]
> We are already at C++11, so if auto was used in the old way, we have a bug.
> 
> As for me, the most interesting thing is modules (C++20), which can potentially reduce compilation time. Though that's less relevant ever since I bought a 16 core machine.
> 
> I did however write more elegant code somewhere using C++17 and reverted because our CICD rejected it.

@royfalk Yes, that particular incident was a pain, wasn't it? Do I recall correctly that Ubuntu Xenial and Debian Stretch were the only builds that rejected that code?

**Comment by stephengtuggy** [comment]
Xenial is nearing end of support. We may be able to drop it soon. Perhaps we could do the same with Stretch.

**Comment by stephengtuggy** [comment]
> C++11 may have better support on OSX and windows to bring in more players -- the only thing really valuable is from C++17 (option, variant classes)---the rest (such as std::span) can come in via various header-only libraries that will defer to C++20, if available (eg https://github.com/martinmoene/span-lite/tree/master/include/nonstd )

@danielrh Oh, good point. I believe our current Windows build environment supports up to C++17. But I have no idea wrt macOS. It's probably best to play it safe with C++11 until we can confirm that that's working on Mac.
